lib/puppet/type/cisco_object_group_entry.rb

```diff
@@ -0,0 +1,138 @@
+# limitations under the License.
+
+require 'ipaddr'
+begin
+  require 'puppet_x/cisco/cmnutils'
+rescue LoadError # seen on master, not on agent
+  # See longstanding Puppet issues #4248, #7316, #14073, #14149, etc. Ugh.
+  require File.expand_path(File.join(File.dirname(__FILE__), '..', '..',
+                                     'puppet_x', 'cisco', 'cmnutils.rb'))
+end
+
+Puppet::Type.newtype(:cisco_object_group_entry) do
+  @doc = "Manages ObjectGroupEntry configuration.
+
+  ~~~puppet
+  cisco_object_group_entry { '<afi> <type> <grp_name> <seqno>':
+    ..attributes..
+  }
+  ~~~
+
+  `<afi> <type> <grp_name> <seqno>` is the title of the object_group_entry resource.
+
+  Example:
+
+  ~~~puppet
+    cisco_object_group_entry { 'ipv4 address my_obj_grp_entry1 10':
+      address                               => '1.2.3.4 2.3.4.5',
+    }
+    cisco_object_group_entry { 'ipv6 address my_obj_grp_entry2 10':
+      address                               => '2000::1/64',
+    }
+    cisco_object_group_entry { 'ipv4 port my_obj_grp_entry3 20':
+        port                              => 'eq 40',
+    }
+    cisco_object_group_entry { 'ipv4 port my_obj_grp_entry4 30':
+        port                              => 'range 40 100',
+    }
+  ~~~
+  "
+
+  ###################
+  # Resource Naming #
+  ###################
+  # Parse out the title to fill in the attributes in these patterns. These
+  # attributes can be overwritten later.
+
+  def self.title_patterns
+    identity = ->(x) { x }
+    [
+      [
+        /^(ipv4|ipv6)\s+(\S+)\s+(\S+)\s+(\d+)$/,
+        [
+          [:afi, identity],
+          [:type, identity],
+          [:grp_name, identity],
+          [:seqno, identity],
+        ],
+      ]
+    ]
+  end
+
+  ##############
+  # Parameters #
+  ##############
+
+  ensurable
+
+  # Overwrites the name method which by default returns only
+  # self[:name].
+  def name
+    "#{self[:afi]} #{self[:type]} #{self[:grp_name]} #{self[:seqno]}"
+  end
+
+  # Only needed to satisfy name parameter.
+  newparam(:name) do
+  end
+
+  newparam(:afi, namevar: true) do
+    desc 'The Address-Family Identifier (ipv4|ipv6).'
+    newvalues(:ipv4, :ipv6)
+  end
+
+  newparam(:type, namevar: true) do
+    desc 'Type of the object_group instance.'
+    newvalues(:address, :port)
+  end
+
+  newparam(:grp_name, namevar: true) do
+    desc 'Object Group Entry name'
+  end
+
+  newparam(:seqno, namevar: true) do
+    desc 'Sequence number of the entry'
+  end
+
+  ##############
+  # Properties #
+  ##############
+
+  newproperty(:address) do
+    desc 'Address to match against. Valid values are an IP'\
+         ' address/prefix_len, IP Address and wildcard, host and '\
+         'host address'
+
+    validate do |address|
+      fail 'address must be ip address/prefix_len,'\
+           ' address/prefix_len, IP Address and wildcard, host and host address' unless
+        %r{host \S+|\S+\/\d+|\S+ [:\.0-9a-fA-F]+}.match(address)
```

Comment:
Might be better to use `PuppetX::Cisco::Utils.process_network_mask()`?

https://github.com/cisco/cisco-network-puppet-module/blob/develop/lib/puppet/type/cisco_bgp_neighbor.rb#L174